src/cargo/ops/fix.rs

```diff
@@ -193,6 +200,114 @@ fn check_version_control(config: &Config, opts: &FixOptions) -> CargoResult<()>
     );
 }
 
+fn check_resolver_change(ws: &Workspace<'_>, opts: &FixOptions) -> CargoResult<()> {
+    let root = ws.root_maybe();
+    match root {
+        MaybePackage::Package(root_pkg) => {
+            if root_pkg.manifest().resolve_behavior().is_some() {
+                // If explicitly specified by the user, no need to check.
+                return Ok(());
+            }
+            // Only trigger if updating the root package from 2018.
+            let pkgs = opts.compile_opts.spec.get_packages(ws)?;
+            if !pkgs.iter().any(|&pkg| pkg == root_pkg) {
+                // The root is not being migrated.
+                return Ok(());
+            }
+            if root_pkg.manifest().edition() != Edition::Edition2018 {
+                // V1 to V2 only happens on 2018 to 2021.
+                return Ok(());
+            }
+        }
+        MaybePackage::Virtual(_vm) => {
+            // Virtual workspaces don't have a global edition to set (yet).
+            return Ok(());
+        }
+    }
+    if !opts.compile_opts.filter.is_all_targets() {
+        // When migrating specific targets, we can't know if the user intends
+        // to set the global edition at this time or not. Conservatively
+        // assume the user will figure things out.
+        return Ok(());
+    }
+    // 2018 without `resolver` set must be V1
+    assert_eq!(ws.resolve_behavior(), ResolveBehavior::V1);
+    let specs = opts.compile_opts.spec.to_package_id_specs(ws)?;
+    let resolve_opts = ResolveOpts::new(
+        /*dev_deps*/ true,
+        RequestedFeatures::from_command_line(
+            &opts.compile_opts.features,
+            opts.compile_opts.all_features,
+            !opts.compile_opts.no_default_features,
+        ),
```

Comment:
Drive-by comment not related to your PR, but this seems like it's ripe for a `opts.copmile_opts.requested_features()` method